client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -60,43 +66,85 @@ private StreamInfo getStreamInfo(Stream stream) {
     }
 
     @Override
-    public Iterator<SegmentInfo> listSegments(Stream stream) {
-        return listSegments(stream, new Date(0L));
+    public StreamSegmentInfo getSegments(Stream stream) {
+        return listSegments(stream, Optional.empty(), Optional.empty());
     }
 
-    private Iterator<SegmentInfo> listSegments(Stream stream, Date from) {
-        // modify iteration above but starting with a timestamp and ending with a break
-        Map<Segment, Long> segments = getAndHandleExceptions(controller.getSegmentsAtTime(new StreamImpl(stream.getScope(),
-                                                                                                         stream.getStreamName()),
-                                                                                          from.getTime()),
-                                                             RuntimeException::new);
+    @Override
+    public StreamSegmentInfo getSegments(final Stream stream, StreamCut fromStreamCut, StreamCut toStreamCut) {
+        Preconditions.checkNotNull(stream, "stream");
+        return listSegments(stream, Optional.ofNullable(fromStreamCut), Optional.ofNullable(toStreamCut));
+    }
+
+    @Override
+    public <T> SegmentIterator<T> readSegment(final SegmentMetadata segment, final Serializer<T> deserializer) {
+        return new SegmentIteratorImpl<>(inputStreamFactory, segment.getSegment(), deserializer,
+                segment.getStartOffset(), segment.getEndOffset());
+    }
+
+    private StreamSegmentInfo listSegments(final Stream stream, final Optional<StreamCut> startStreamCut,
```

Comment:
nit: use startStreamCut and endStreamCut or fromStreamCut and toStreamCut
`start` and `to` are odd.

Reply:
done.